storage/cloud: Make it possible to disable external http storage

miretskiy opened the pull request:
Informs #44320

Make it possible to disable external http storage.

This change enables the operators to disable access
 to external http servers as well as custom http endpoints
 overrides for cloud storage implementations.

Release note (security update): Make it possible for operators
to disable external http access when performing certain
operations (BACKUP, IMPORT, etc).

The external http access, as well as custom http endpoints, are
disabled by providing an --external-io-disable-http flag.

This flag provides a light weight option to disable http external
access in the environments where running a full fledged proxy
server may not be feasible.  If running a proxy service is
acceptible, operators may choose to start cockroach binary
specifying HTTP(s)_PROXY environment setting.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44900)
<!-- Reviewable:end -->

dt commented on pkg/cli/flags.go:
```diff
@@ -360,6 +403,14 @@ func init() {
 		BoolFlag(f, &startCtx.serverInsecure, cliflags.ServerInsecure, startCtx.serverInsecure)
 
+		// Enable/disable various external storage endpoints.
+		// TODO(yevgeniy): currently we enable all schemas by default, but we should probably
```
I'm not sure about this point about defaults -- I think we want s3/gcs/etc to "just work" out of the box, and I think the default is probably to have most of them on. Indeed, I'm sort of inclined to invert the struct fields to be `DisableFoo` form so the zero-value is everything on, since flipping one off, like HTTP, is expected to be the exception, not the rule?

I donno, maybe @mwang1026 has thoughts

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44900)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mwang1026)

___
*[pkg/cli/flags.go, line 407 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44900#-M-bmb4Hcujkftxl6rCg-r1-407:-M-ppxwp3D6O6iqxkS__:b-g20jlm) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a5f1951189ce36f8d5670e9b5919e92473988f9/pkg/cli/flags.go#L407)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I'm not sure about this point about defaults -- I think we want s3/gcs/etc to "just work" out of the box, and I think the default is probably to have most of them on. Indeed, I'm sort of inclined to invert the struct fields to be `DisableFoo` form so the zero-value is everything on, since flipping one off, like HTTP, is expected to be the exception, not the rule?

I donno, maybe @mwang1026 has thoughts
</blockquote></details>

I started off with just that... Disable http.
However, I kinda feel like if you want to be paranoid, you might want to disable most of them.
Or at least want to be able to disable any of them...


<!-- Sent from Reviewable.io -->

Commit:
This change enables the operators to disable access
to external http servers as well as custom http endpoints
overrides for cloud storage implementations.

Informs #44320

Release note (security update): Make it possible for operators
to disable external http access when performing certain
operations (BACKUP, IMPORT, etc).

The external http access, as well as custom http endpoints, are
disabled by providing an --external-io-disable-http flag.

This flag provides a light weight option to disable http external
access in the environments where running a full fledged proxy
server may not be feasible.  If running a proxy service is
acceptible, operators may choose to start cockroach binary
specifying HTTP(s)_PROXY environment setting.

mwang1026 commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44900)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mwang1026)
> 
> _[pkg/cli/flags.go, line 407 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44900#-M-bmb4Hcujkftxl6rCg-r1-407:-M-ppxwp3D6O6iqxkS__:b-g20jlm) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a5f1951189ce36f8d5670e9b5919e92473988f9/pkg/cli/flags.go#L407)):_
> 
> Previously, dt (David Taylor) wrote…
> > I'm not sure about this point about defaults -- I think we want s3/gcs/etc to "just work" out of the box, and I think the default is probably to have most of them on. Indeed, I'm sort of inclined to invert the struct fields to be `DisableFoo` form so the zero-value is everything on, since flipping one off, like HTTP, is expected to be the exception, not the rule?
> > I donno, maybe @mwang1026 has thoughts
> 
> I started off with just that... Disable http.
> However, I kinda feel like if you want to be paranoid, you might want to disable most of them.
> Or at least want to be able to disable any of them...

If the action is to disable that makes sense to me for it to be `DisableFoo`. If you are paranoid you're probably willing to do the work to disable.

dt commented on pkg/storage/cloud/s3_storage.go:
```diff
@@ -55,14 +56,20 @@ func s3QueryParams(conf *roachpb.ExternalStorage_S3) string {
 	config := conf.Keys()
 	if conf.Endpoint != "" {
+		if ioConf.DisableHTTP {
+			return nil, errors.New("Custom endpoints disallowed for s3")
```
nit: imo this error would be a bit more helpful if it explained why -- that the http disallow flag was the reason.

knz replied:
additionally, error messages do not start with a capital - they are not sentences and can be embedded in larger sentences

miretskiy replied:
Done.

knz commented:
The commit message is a bit messed up:

- the first line should be the commit message title
- the "Informs" line should be at the beginning, otherwise it becomes part of the release note.

Check this: https://wiki.crdb.io/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages

miretskiy commented:
bors r+
